File: CRM.Web/src/api/financeReceivable.ts
import apiClient from './client'
import type { FinanceCustomerAdvanceBalance } from './financeCustomerAdvance'
import { toExportQueryString } from '@/utils/exportFileName'

export interface FinanceReceivable {
  id: string
  receivableCode?: string
  stockOutId: string
  stockOutCode: string
  sellOrderId: string
  sellOrderCode?: string
  sellOrderItemId: string
  customerId: string
  customerName?: string
  customerEnglishName?: string
  customerCode?: string
  salesUserId?: string
  salesUserName?: string
  pn?: string
  brand?: string
  outboundQty: number
  unitPrice: number
  currency: number
  amount: number
  verifiedDone: number
  verifiedToBe: number
  verificationStatus: number
  stockOutDate?: string
  createTime?: string
}

export interface FinanceReceivableWriteOffDetailItem {
  id: string
  amount: number
  writeOffSource: number
  createTime?: string
  financeReceiptId?: string
  financeReceiptItemId?: string
  financeReceiptCode?: string
  operatorUserId?: string
  operatorUserName?: string
  remark?: string
}

export interface FinanceReceiptItemWriteOffCandidate {
  item: {
    id: string
    financeReceiptId: string
    receiptAmount: number
    receiptConvertAmount?: number
    verifiedAmount: number
    advancePoolAmount?: number
    verificationStatus: number
    receiptPurpose?: number
    advanceSellOrderId?: string
    sellOrderId?: string
    pn?: string
    brand?: string
    remark?: string
  }
  financeReceiptCode: string
  receiptStatus: number
  remainingAmount: number
  receiptPurpose?: number
  advanceSellOrderId?: string
  receiptDate?: string
  receiptAmount: number
  receiptCurrency: number
  receiptMode: number
  remark?: string
}

export interface FinanceReceivableWriteOffCandidateRow {
  id: string
  receivableCode?: string
  stockOutId: string
  stockOutCode: string
  sellOrderId: string
  sellOrderCode?: string
  sellOrderItemId: string
  customerId: string
  customerName?: string
  customerEnglishName?: string
  salesUserId?: string
  salesUserName?: string
  pn?: string
  brand?: string
  outboundQty: number
  unitPrice: number
  currency: number
  amount: number
  verifiedDone: number
  verifiedToBe: number
  verificationStatus: number
  stockOutDate?: string
  freightForwarderOrderNo?: string
  stockInCode?: string
}

export interface FinanceWriteOffCustomerCurrencyTotal {
  currency: number
  amount: number
}

export interface FinanceWriteOffCustomerSummary {
  customerId: string
  customerName?: string
  customerEnglishName?: string
  customerCode?: string
  salesUserId?: string
  salesUserName?: string
  pendingWriteOffTotal: number
  currency?: number | null
  isMultiCurrency: boolean
  currencyTotals: FinanceWriteOffCustomerCurrencyTotal[]
  pendingReceiptItemCount: number
  earliestReceiptDate?: string | null
  latestReceiptDate?: string | null
  /** 该客户该币别是否存在未清应收（与右栏口径一致） */
  hasOpenReceivable?: boolean
}

export interface FinanceReceivableWriteOffCandidates {
  receiptItems: FinanceReceiptItemWriteOffCandidate[]
  receivables: FinanceReceivableWriteOffCandidateRow[]
  advanceBalances: FinanceCustomerAdvanceBalance[]
}

export interface FinanceReceivableWriteOffAllocation {
  financeReceiptItemId: string
  financeReceivableId: string
  amount: number
}

export interface FinanceAdvancePoolAllocation {
  financeReceivableId: string
  amount: number
  advanceSellOrderId?: string
}

export interface FinanceReceivableWriteOffSoMismatch {
  financeReceivableId: string
  advanceSellOrderId?: string
  receivableSellOrderId?: string
  message?: string
}

export interface FinanceReceivableWriteOffResult {
  applied: boolean
  requiresSoMismatchConfirm: boolean
  soMismatches: FinanceReceivableWriteOffSoMismatch[]
}

export interface CreditReceiptItemRemainderToPoolResult {
  creditedAmount: number
  remainingAfter: number
}

export interface FinanceReceivableWriteOffLedgerItem {
  id: string
  amount: number
  writeOffSource: number
  createTime?: string
  financeReceiptId?: string
  financeReceiptItemId?: string
  financeReceiptCode?: string
  financeReceivableId: string
  receivableCode?: string
  stockOutId?: string
  stockOutCode?: string
  sellOrderId?: string
  sellOrderCode?: string
  customerId: string
  customerName?: string
  customerEnglishName?: string
  pn?: string
  brand?: string
  currency: number
  operatorUserId?: string
  operatorUserName?: string
  remark?: string
}

export interface FinanceReceivableWriteOffLedgerPage {
  items: FinanceReceivableWriteOffLedgerItem[]
  total: number
  page: number
  pageSize: number
}

export interface FinanceReceivableWriteOffRequest {
  allocations: FinanceReceivableWriteOffAllocation[]
  advancePoolAllocations?: FinanceAdvancePoolAllocation[]
  confirmSoMismatch?: boolean
}

const RECEIVABLE_BASE = '/api/v1/finance/receivables'
const WRITE_OFF_BASE = '/api/v1/finance/receivable-write-offs'

export const financeReceivableApi = {
  getPaged: (params: Record<string, unknown>) =>
    apiClient.get<{ items: FinanceReceivable[]; total: number; page: number; pageSize: number }>(
      RECEIVABLE_BASE,
      { params }
    ),
  exportList: (params?: Record<string, unknown>) => {
    const q = toExportQueryString(params)
    return apiClient.getBlob(q ? `${RECEIVABLE_BASE}/export?${q}` : `${RECEIVABLE_BASE}/export`)
  },
  getById: (id: string) => apiClient.get<FinanceReceivable>(`${RECEIVABLE_BASE}/${encodeURIComponent(id)}`),
  getWriteOffs: (id: string) =>
    apiClient.get<FinanceReceivableWriteOffDetailItem[]>(`${RECEIVABLE_BASE}/${encodeURIComponent(id)}/write-offs`),
  getWriteOffCustomerSummaries: (keyword?: string) =>
    apiClient.get<FinanceWriteOffCustomerSummary[]>(`${WRITE_OFF_BASE}/customer-summaries`, {
      params: keyword?.trim() ? { keyword: keyword.trim() } : undefined
    }),
  getWriteOffLedger: (params: { keyword?: string; page?: number; pageSize?: number }) =>
    apiClient.get<FinanceReceivableWriteOffLedgerPage>(`${WRITE_OFF_BASE}/ledger`, { params }),
  getWriteOffCandidates: (customerId: string) =>
    apiClient.get<FinanceReceivableWriteOffCandidates>(`${WRITE_OFF_BASE}/candidates`, {
      params: { customerId }
    }),
  applyWriteOff: (payload: FinanceReceivableWriteOffRequest) =>
    apiClient.post<FinanceReceivableWriteOffResult>(WRITE_OFF_BASE, payload),
  creditReceiptItemRemainderToAdvancePool: (receiptItemId: string, amount?: number) =>
    apiClient.post<CreditReceiptItemRemainderToPoolResult>(
      `${WRITE_OFF_BASE}/receipt-items/${encodeURIComponent(receiptItemId)}/credit-to-advance-pool`,
      amount != null && amount > 0 ? { amount } : {}
    )
}

File: CRM.Core/Models/Finance/FinanceReceivable.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using CRM.Core.Interfaces;

namespace CRM.Core.Models.Finance;

/// <summary>应收款（销售出库批次维度，一出库单头一行）。</summary>
[Table("finance_receivable")]
public class FinanceReceivable : BaseGuidEntity, ISoftDeletable
{
    [Key]
    [StringLength(36)]
    [Column("FinanceReceivableId")]
    public override string Id { get; set; } = Guid.NewGuid().ToString();

    [StringLength(16)]
    public string? ReceivableCode { get; set; }

    [Required]
    [StringLength(36)]
    [Column("stock_out_id")]
    public string StockOutId { get; set; } = string.Empty;

    [Required]
    [StringLength(32)]
    public string StockOutCode { get; set; } = string.Empty;

    [Required]
    [StringLength(36)]
    [Column("sell_order_id")]
    public string SellOrderId { get; set; } = string.Empty;

    [StringLength(32)]
    [Column("sell_order_code")]
    public string? SellOrderCode { get; set; }

    [Required]
    [StringLength(36)]
    [Column("sell_order_item_id")]
    public string SellOrderItemId { get; set; } = string.Empty;

    [Required]
    [StringLength(36)]
    [Column("customer_id")]
    public string CustomerId { get; set; } = string.Empty;

    [StringLength(200)]
    [Column("customer_name")]
    public string? CustomerName { get; set; }

    /// <summary>客户英文名（详情/列表 API 填充，不落库）</summary>
    [NotMapped]
    public string? CustomerEnglishName { get; set; }

    [StringLength(36)]
    [Column("sales_user_id")]
    public string? SalesUserId { get; set; }

    [StringLength(200)]
    public string? PN { get; set; }

    [StringLength(200)]
    public string? Brand { get; set; }

    [Column("outbound_qty", TypeName = "numeric(18,4)")]
    public decimal OutboundQty { get; set; }

    [Column("unit_price", TypeName = "numeric(18,6)")]
    public decimal UnitPrice { get; set; }

    /// <summary>币别 1=RMB 2=USD 3=EUR</summary>
    public short Currency { get; set; } = 1;

    [Column(TypeName = "numeric(18,2)")]
    public decimal Amount { get; set; }

    [Column("verified_done", TypeName = "numeric(18,2)")]
    public decimal VerifiedDone { get; set; }

    [Column("verified_to_be", TypeName = "numeric(18,2)")]
    public decimal VerifiedToBe { get; set; }

    /// <summary>核销状态 0未核销 1部分核销 2核销完成</summary>
    [Column("verification_status")]
    public short VerificationStatus { get; set; }

    /// <summary>已匹配开票金额（票↔应收）</summary>
    [Column("invoice_match_done", TypeName = "numeric(18,2)")]
    public decimal InvoiceMatchDone { get; set; }

    /// <summary>待匹配开票金额</summary>
    [Column("invoice_match_to_be", TypeName = "numeric(18,2)")]
    public decimal InvoiceMatchToBe { get; set; }

    /// <summary>开票匹配状态 0未匹配 1部分 2完成</summary>
    [Column("invoice_match_status")]
    public short InvoiceMatchStatus { get; set; }

    /// <summary>开票匹配币别（可选冗余）</summary>
    [Column("invoice_match_currency")]
    public short? InvoiceMatchCurrency { get; set; }

    [Column("stock_out_date")]
    public DateTime? StockOutDate { get; set; }

    [Column("is_deleted")]
    public bool IsDeleted { get; set; }

    [StringLength(36)]
    public string? CreateByUserId { get; set; }

    [StringLength(36)]
    public string? ModifyByUserId { get; set; }
}
